### PR TITLE
Rework the simplifier to use ConstantInterval for bounds

### PR DESCRIPTION
**Pull request by abadams**
Now that ConstantInterval is more powerful, this PR reworks the simplifier to use it for bounds information. This is not quite just a refactor PR, because it also tracks bounds for many more types, resulting in better simplification of e.g. fixed-point expressions.

I also moved the simplify fuzzer back to being a correctness test (see #8171)

Also includes bug-fixes #8220 and #8221, because those would otherwise break this PR. There are no files in common, so if you want to review before those are merged, just ignore the associativity and modulus remainder changes.

### CODE REVIEW

File: src/Simplify_Call.cpp [hunk 1]
```diff
@@ -355,14 +360,12 @@ Expr Simplify::visit(const Call *op, ExprInfo *bounds) {
         }
     } else if (op->is_intrinsic(Call::saturating_cast)) {
         internal_assert(op->args.size() == 1);
-        ExprInfo a_bounds;
-        Expr a = mutate(op->args[0], &a_bounds);
-
-        // TODO(rootjalex): We could be intelligent about using a_bounds to remove saturating_casts;
```
Comment (rootjalex) [inline src/Simplify_Call.cpp:361]: Are you removing this TODO because you think it's not worth doing?

File: src/Simplify_Div.cpp [hunk 2]
```diff
@@ -3,112 +3,45 @@
...
+    if (info) {
+        if (op->type.is_int_or_uint()) {
+            info->bounds = a_info.bounds / b_info.bounds;
+            info->alignment = a_info.alignment / b_info.alignment;
+            info->trim_bounds_using_alignment();
+            info->cast_to(op->type);
+
+            // Bounded numerator divided by constantish
+            // denominator can sometimes collapse things to a
+            // constant at this point
```
Comment (rootjalex) [inline src/Simplify_Div.cpp:20]: What is "constantish"?

File: src/Simplify_Div.cpp [hunk 3]
```diff
@@ -3,112 +3,45 @@
...
+                    // we're better off returning an overflow condition than
+                    // a known-wrong value. (Note that no_overflow_int() should
+                    // only be true for signed integers.)
+                    internal_assert(no_overflow_int(op->type));
+                    clear_expr_info(info);
+                    return make_signed_integer_overflow(op->type);
+                }
             }
+        } else {
+            // TODO: division bounds for floating-point division
```
Comment (rootjalex) [inline src/Simplify_Div.cpp:36]: Does this warrant an open issue? We didn't previously support bounds on floating-point division, right?

File: src/Simplify_Call.cpp (hunk 1 above)
Comment (abadams) [reply src/Simplify_Call.cpp:361]: I think converting a saturating cast to a cast is likely to mess up instruction selection later

File: src/Simplify_Div.cpp (hunk 3 above)
Comment (abadams) [reply src/Simplify_Div.cpp:36]: That's correct. Main only supports int32/64 constant bounds.

I don't think it's worth an issue. This is not something I'd suggest someone to work on. It's more of a placeholder to say "hey, if we need this, here's where to put the logic"

File: src/Simplify_Div.cpp (hunk 2 above)
Comment (abadams) [reply src/Simplify_Div.cpp:20]: This is just code motion (increased indent level), but there are cases where the denominator doesn't strictly need to be a constant that this handles, e.g. [3,9] / [15,23] = 0

File: src/Simplify_Call.cpp (hunk 1 above)
Comment (rootjalex) [reply src/Simplify_Call.cpp:361]: Ah, for ISAs that only support saturating casts? Maybe we need #7805 - style instruction selection to rewrite casts to saturating casts on those platforms

File: src/Simplify_Div.cpp (hunk 2 above)
Comment (rootjalex) [reply src/Simplify_Div.cpp:20]: I think it would be useful to clarify that in the comment, maybe just adding that example

File: src/Simplify_Max.cpp [hunk 4]
```diff
@@ -3,52 +3,51 @@
...
-            bounds->min = b_bounds.min;
-        }
-        bounds->alignment = ModulusRemainder::unify(a_bounds.alignment, b_bounds.alignment);
-        bounds->trim_bounds_using_alignment();
+Expr Simplify::visit(const Max *op, ExprInfo *info) {
+    ExprInfo a_info, b_info;
+    Expr a = mutate(op->a, &a_info);
+    Expr b = mutate(op->b, &b_info);
+
+    // debug(0) << "max(" << a_info.bounds << " " << b_info.bounds << ")\n";
```
Comment (rootjalex) [inline src/Simplify_Max.cpp:11]: Debugging statement left in (and a few lines below)

File: src/Simplify_Select.cpp [hunk 5]
```diff
@@ -3,20 +3,23 @@
...
-        bounds->alignment = ModulusRemainder::unify(t_bounds.alignment, f_bounds.alignment);
-        bounds->trim_bounds_using_alignment();
+    Expr true_value = mutate(op->true_value, &t_info);
+    Expr false_value = mutate(op->false_value, &f_info);
+
+    if (info) {
+        info->bounds = ConstantInterval::make_union(t_info.bounds, f_info.bounds);
+        info->alignment = ModulusRemainder::unify(t_info.alignment, f_info.alignment);
+        /*
+        debug(0) << t_info.bounds << " " << t_info.alignment << "\n"
```
Comment (rootjalex) [inline src/Simplify_Select.cpp:17]: debugging code left in (and below)

File: src/Simplify_Stmts.cpp [hunk 6]
```diff
@@ -218,23 +218,36 @@ Stmt Simplify::visit(const For *op) {
...
-        ScopedFact fact = scoped_truth(0 < new_extent);
+        ScopedFact fact = scoped_truth(extent_positive);
         new_body = mutate(op->body);
     }
     if (in_unreachable) {
-        if (extent_bounds.min_defined && extent_bounds.min >= 1) {
-            // If we know the loop executes once, the code that runs this loop is unreachable.
+        if (extent_info.bounds > 0) {
+            // If we know the loop executes at least once, the code that runs
+            // this loop is unreachable.
```
Comment (rootjalex) [inline src/Simplify_Stmts.cpp:250]: Not sure if I'm just struggling to parse this, but this comment doesn't make sense to me (the before or edit). If the lop executes, then the code inside the loop is certainly *reachable*, no?

File: test/correctness/fuse.cpp [hunk 7]
```diff
@@ -72,7 +72,7 @@ int main(int argc, char **argv) {
         Var xy("xy");
         f.compute_root()
             .fuse(x, y, xy)
-            .vectorize(xy, 16);
+            .vectorize(xy, 16, TailStrategy::RoundUp);
```
Comment (rootjalex) [inline test/correctness/fuse.cpp:75]: Is this change related to the PR?

File: test/correctness/fuzz_simplify.cpp [hunk 8]
```diff
@@ -184,9 +180,9 @@ Expr random_expr(FuzzedDataProvider &fdp, Type t, int depth, bool overflow_undef
                 make_absd,
             };
 
-            Expr a = random_expr(fdp, t, depth, overflow_undef);
-            Expr b = random_expr(fdp, t, depth, overflow_undef);
-            return fdp.PickValueInArray(make_bin_op)(a, b);
+            Expr a = random_expr(rng, t, depth, overflow_undef);
+            Expr b = random_expr(rng, t, depth, overflow_undef);
+            return make_bin_op[rng() % 8](a, b);
```
Comment (rootjalex) [inline test/correctness/fuzz_simplify.cpp:185]: Using `rng() % (sizeof(make_bin_op) / sizeof(make_bin_op[0])` here and below might be better imo, less magic constants.

File: test/correctness/fuzz_simplify.cpp [hunk 9]
```diff
@@ -266,96 +262,62 @@ bool test_expression(FuzzedDataProvider &fdp, Expr test, int samples) {
     return true;
 }
 
-// These are here to enable copy of failed output expressions
```
Comment (rootjalex) [inline test/correctness/fuzz_simplify.cpp:269]: Is there a reason for removing these? I found them to be useful for debugging when I've been debugging fuzz failures.

**Review by rootjalex — APPROVED** [review]
Lgtm with nits

File: test/correctness/fuzz_simplify.cpp (hunk 9 above)
Comment (abadams) [reply test/correctness/fuzz_simplify.cpp:269]: They're a lot less necessary now that you can just use the same seed, and now that all subexpressions are tested too, so it's probably already reduced.

File: test/correctness/fuse.cpp (hunk 7 above)
Comment (abadams) [reply test/correctness/fuse.cpp:75]: Yes, the stronger simplification injected a new mod operation, which wasn't the mod operation we were trying to assert didn't exist.

File: test/correctness/fuzz_simplify.cpp (hunk 8 above)
Comment (abadams) [reply test/correctness/fuzz_simplify.cpp]: I just added a random_choice function to remove all these mods

File: src/Simplify_Stmts.cpp (hunk 6 above)
Comment (abadams) [reply src/Simplify_Stmts.cpp]: Please take another look and tell me if it's clearer now. I also found a bug here related to not popping the expr info for the loop var, so the code has been reshuffled a little.

Comment (rootjalex) [reply src/Simplify_Stmts.cpp]: Change lgtm! Thanks

**Comment by abadams** [comment]
Failures unrelated. Merging.

File: test/correctness/fuzz_simplify.cpp [hunk 10]
```diff
@@ -25,65 +23,78 @@ std::string fuzz_var(int i) {
     return std::string(1, 'a' + i);
 }
 
-Expr random_var(FuzzedDataProvider &fdp) {
-    int fuzz_count = fdp.ConsumeIntegralInRange<int>(0, fuzz_var_count - 1);
-    return Variable::make(Int(0), fuzz_var(fuzz_count));
+Expr random_var(std::mt19937 &rng, Type t) {
+    int fuzz_count = rng() % (fuzz_var_count - 1);
+    return cast(t, Variable::make(Int(32), fuzz_var(fuzz_count)));
 }
```
Comment (alexreinking) [inline test/correctness/fuzz_simplify.cpp:29]: This technically introduced a bug since `ConsumeIntegralInRange` is inclusive on both sides.
